### PR TITLE
[state-sync] Optimize weight distribution calculation for peer scoring

### PR DESCRIPTION
**Pull request by sunmilee**
## Motivation

WeightedIndex library can be O(log(n)) if the weighted index vector does not need to be updated when choosing a peer. This diff adds logic to only update the weighted index vector when score changes, and store that vector to be used when picking peer, rather than calculating the distribution dynamically when picking.

## Test Plan

unit test

### CODE REVIEW

File: state_synchronizer/src/peer_manager.rs [hunk 1]
```diff
@@ -39,17 +39,20 @@ pub enum PeerScoreUpdateType {
 pub struct PeerManager {
     peers: HashMap<PeerId, PeerInfo>,
     network_senders: HashMap<PeerId, StateSynchronizerSender>,
+    weighted_index: Option<WeightedIndex<f64>>,
 }
 
 impl PeerManager {
     pub fn new(peer_ids: Vec<PeerId>) -> Self {
-        let peers = peer_ids
+        let peers: HashMap<PeerId, PeerInfo> = peer_ids
```
Comment (bothra90) [inline state_synchronizer/src/peer_manager.rs]: nit: do you need the explicit type annotation?

File: state_synchronizer/src/peer_manager.rs [hunk 2]
```diff
@@ -94,28 +101,51 @@ impl PeerManager {
         if let Some(peer_info) = self.peers.get_mut(peer_id) {
             match update_type {
                 PeerScoreUpdateType::Success => {
-                    let new_score = peer_info.score + 1.0;
-                    peer_info.score = new_score.min(MAX_SCORE);
+                    if peer_info.score < MAX_SCORE {
+                        let new_score = peer_info.score + 1.0;
+                        peer_info.score = new_score.min(MAX_SCORE);
+                        self.weighted_index = self.compute_weighted_index();
```
Comment (bothra90) [inline state_synchronizer/src/peer_manager.rs]: do this outside the match?

File: state_synchronizer/src/peer_manager.rs [hunk 3]
```diff
@@ -94,28 +101,51 @@ impl PeerManager {
...
                 .iter()
                 .map(|(_, peer_info)| peer_info.score)
                 .collect();
+            match WeightedIndex::new(&weights) {
+                Ok(weighted_index) => {
+                    return Some(weighted_index);
+                }
+                Err(e) => {
+                    error!(
+                        "[state sync] (pick_peer) invalid weighted index distribution, {:?}",
```
Comment (bothra90) [inline state_synchronizer/src/peer_manager.rs]: should we say something like "Failed to create weighted index"?

File: state_synchronizer/src/peer_manager.rs [hunk 4]
```diff
@@ -94,28 +101,51 @@ impl PeerManager {
...
+        None
+    }
+
+    pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
+        let active_peers = self.get_active_upstream_peers();
+        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
 
-            if let Ok(weighted_index) = WeightedIndex::new(&weights) {
+        if !active_peers.is_empty() {
+            if let Some(weighted_index) = self.weighted_index.clone() {
```
Comment (bothra90) [inline state_synchronizer/src/peer_manager.rs]: can we sample without cloning the weighted index?

File: state_synchronizer/src/peer_manager.rs [hunk 5]
```diff
@@ -76,6 +79,10 @@ impl PeerManager {
             self.peers
                 .insert(peer_id, PeerInfo::new(true, false, MAX_SCORE));
         }
+
+        if self.weighted_index.is_none() {
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: if it's a new peer, but not yet included in index, index should be rebuilded

also disable_peer might effect score

File: state_synchronizer/src/peer_manager.rs (hunk 2 above)
Comment (phoenix-antigravity) [reply state_synchronizer/src/peer_manager.rs]: yeah, you can just get old score. Perform update and compare it

File: state_synchronizer/src/peer_manager.rs (hunk 4 above)
Comment (phoenix-antigravity) [reply state_synchronizer/src/peer_manager.rs]: yeah, isn't clone diminishes whole optimization?

File: state_synchronizer/src/peer_manager.rs [hunk 6]
```diff
@@ -76,6 +79,9 @@ impl PeerManager {
             self.peers
                 .insert(peer_id, PeerInfo::new(true, false, MAX_SCORE));
         }
+
+        self.weighted_index = self.compute_weighted_index();
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: nit api change: can we make `compute_weighted_index` to be avoid and just alter internal state. Cause it's called with assignment everywhere anyway

Comment (phoenix-antigravity) [reply state_synchronizer/src/peer_manager.rs]: also should it be also called for set_peers?

File: state_synchronizer/src/peer_manager.rs [hunk 7]
```diff
@@ -92,30 +99,57 @@ impl PeerManager {
 
     pub fn update_score(&mut self, peer_id: &PeerId, update_type: PeerScoreUpdateType) {
         if let Some(peer_info) = self.peers.get_mut(peer_id) {
+            let mut is_update = false;
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: why not just do `let old_score =  peer_info.score`
and then later compare that they are not equal?

File: state_synchronizer/src/peer_manager.rs (hunk 6 above)
Comment (sunmilee) [reply state_synchronizer/src/peer_manager.rs]: I guess it should be? but should only be updated if any of the new peers are alive and upstream

File: state_synchronizer/src/peer_manager.rs [hunk 8]
```diff
@@ -92,30 +108,54 @@ impl PeerManager {
...
                 PeerScoreUpdateType::InvalidChunk => {
-                    let new_score = peer_info.score * 0.8;
-                    peer_info.score = new_score.max(MIN_SCORE);
+                    if peer_info.score > MIN_SCORE {
+                        let new_score = peer_info.score * 0.8;
+                        peer_info.score = new_score.max(MIN_SCORE);
+                    }
                 }
             }
+            if (old_score - peer_info.score).abs() > std::f64::EPSILON {
```
Comment (sunmilee) [inline state_synchronizer/src/peer_manager.rs]: addressing clippy error     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider comparing them within some error: `(old_score - peer_info.score).abs() > error`

File: state_synchronizer/src/peer_manager.rs [hunk 9]
```diff
@@ -47,17 +48,29 @@ impl PeerManager {
...
+        let mut update_weighted_index = false;
         for (peer_id, info) in self.peers.iter_mut() {
-            info.is_upstream = new_peer_ids.contains(peer_id);
+            let new_is_upstream = new_peer_ids.contains(peer_id);
+            if info.is_upstream != new_is_upstream {
+                update_weighted_index = true;
+            }
+            info.is_upstream = new_is_upstream;
         }
+        if update_weighted_index {
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: nit: IMO that's unnecessary optimization, cause set_peers should be relatively rare call (only used by Consensus) and building index is not a perf concern for it. I'd prefer readability here. Just  call compute_weighted_index unconditionally

File: state_synchronizer/src/peer_manager.rs [hunk 10]
```diff
@@ -92,30 +108,54 @@ impl PeerManager {
 
     pub fn update_score(&mut self, peer_id: &PeerId, update_type: PeerScoreUpdateType) {
         if let Some(peer_info) = self.peers.get_mut(peer_id) {
+            let old_score = peer_info.score;
             match update_type {
                 PeerScoreUpdateType::Success => {
-                    let new_score = peer_info.score + 1.0;
-                    peer_info.score = new_score.min(MAX_SCORE);
+                    if peer_info.score < MAX_SCORE {
```
Comment (phoenix-antigravity) [inline state_synchronizer/src/peer_manager.rs]: nit: why keep those if statements?

File: state_synchronizer/src/peer_manager.rs [hunk 11]
```diff
@@ -102,20 +108,39 @@ impl PeerManager {
...
+            }
+        }
+    }
+
+    pub fn pick_peer(&self) -> Option<(PeerId, StateSynchronizerSender)> {
+        let active_peers = self.get_active_upstream_peers();
+        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
 
-            if let Ok(weighted_index) = WeightedIndex::new(&weights) {
+        if !active_peers.is_empty() {
```
Comment (bothra90) [inline state_synchronizer/src/peer_manager.rs]: do we need this check?

Comment (sunmilee) [reply state_synchronizer/src/peer_manager.rs]: I think so because we fetch using index later? `active_peers[weighted_index.sample(&mut rng)]`

Comment (phoenix-antigravity) [reply state_synchronizer/src/peer_manager.rs]: we actually don't need this specific check. But it bothers me a bit that if active_peers and weighted_index will go out of sync, we'd panic on line 145

Comment (sunmilee) [reply state_synchronizer/src/peer_manager.rs]: Hm maybe I should check that `weighted_index.len() == active_peers.len()`? And recompute weighted index if they are out of sync for some reason. That check wouldn't guarantee that each index refers to the correct peer, but it at least wouldn't panic

Comment (phoenix-antigravity) [reply state_synchronizer/src/peer_manager.rs]: just use `.get` api instead of `[..]` call

Comment (bothra90) [reply state_synchronizer/src/peer_manager.rs]: I agree with the concern - if it is expected that `weighted_index.len() == active_peers.len()`, could we wrap the 2 within a common struct? We might then be able to use MIRAI macros to assert the invariant.

Comment (sunmilee) [reply state_synchronizer/src/peer_manager.rs]: Updated the function to use .get and removed active.is_empty check. Think that should be sufficient for now, let me know your thoughts

**Pull request merged**